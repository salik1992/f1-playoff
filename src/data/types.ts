export type Driver = {
    name: string;
    results: readonly (number | null)[];
    position: number;
    points: number;
    pointsFromLastRace: number;
    finishesCount: Record<number, number>;
    isInPlayOff: boolean;
    playOffPoints: number;
    playOffFinishesCount: Record<number, number>;
};

export type Season = {
    year: number;
    races: readonly string[];
    drivers: readonly Driver[];
    awardedPoints: Record<number, number>;
    halfPoints: readonly boolean[];
    fastestLapPoints: readonly string[];
};
