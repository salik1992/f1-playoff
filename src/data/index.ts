export { PATHS } from './paths';
export { loadSeason } from './loader';
export type { Driver, Season } from './types';
