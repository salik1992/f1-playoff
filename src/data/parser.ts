import { compose } from '@salik1992/fun-ts/dist/core';
import { filter, forEach, map } from '@salik1992/fun-ts/dist/array';
import { split } from '@salik1992/fun-ts/dist/string';
import { parseInt } from '@salik1992/fun-ts/dist/math';
import { Season, Driver } from './types';

const getValues = compose(
    filter<string>((_, index) => index > 0),
    split(','),
);

const getResults = compose(
    map<number, null | number>((result) => Number.isNaN(result) ? null : result),
    map(parseInt()),
);

const driverLineToDriver = (line: string): Driver => {
    const [name, ...results] = split(',')(line);
    return {
        name,
        results: getResults(results),
        finishesCount: {},
        position: 0,
        points: 0,
        pointsFromLastRace: 0,
        isInPlayOff: false,
        playOffPoints: 0,
        playOffFinishesCount: {},
    }
}

export const parseSeason = (csv: string): Season => {
    const [
        yearLine,
        pointsLine,
        racesLine,
        halfPointsLine,
        fastestLapPointLine,
        ...driversLines
    ] = compose(filter<string>((_, index) => index !== 5), split('\n'))(csv);
    const year = parseInt()(getValues(yearLine)[0]);
    const awardedPoints: Record<number, number> = {};
    const points = compose(
        map<number, number>((points) => Number.isNaN(points) ? 0 : points),
        map<string, number>(parseInt()),
        getValues,
    )(pointsLine);
    forEach<number>((points, index) => {
        awardedPoints[index + 1] = points
    })(points);
    const races = getValues(racesLine);
    const halfPoints = compose(
        map((text) => text === 'YES'),
        getValues,
    )(halfPointsLine);
    const fastestLapPoints = getValues(fastestLapPointLine);
    const drivers = map(driverLineToDriver)(driversLines);
    return {
        year, awardedPoints, drivers, races, halfPoints, fastestLapPoints,
    }
};
