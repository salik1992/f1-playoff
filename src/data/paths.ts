/// <reference path="paths.d.ts" />

import s1991 from './seasons/1991.csv';
import s1992 from './seasons/1992.csv';
import s1993 from './seasons/1993.csv';
import s1994 from './seasons/1994.csv';
import s1995 from './seasons/1995.csv';
import s1996 from './seasons/1996.csv';
import s1997 from './seasons/1997.csv';
import s1998 from './seasons/1998.csv';
import s1999 from './seasons/1999.csv';
import s2000 from './seasons/2000.csv';
import s2001 from './seasons/2001.csv';
import s2002 from './seasons/2002.csv';
import s2003 from './seasons/2003.csv';
import s2004 from './seasons/2004.csv';
import s2005 from './seasons/2005.csv';
import s2006 from './seasons/2006.csv';
import s2007 from './seasons/2007.csv';
import s2008 from './seasons/2008.csv';
import s2009 from './seasons/2009.csv';
import s2010 from './seasons/2010.csv';
import s2011 from './seasons/2011.csv';
import s2012 from './seasons/2012.csv';
import s2013 from './seasons/2013.csv';
import s2014 from './seasons/2014.csv';
import s2015 from './seasons/2015.csv';
import s2016 from './seasons/2016.csv';
import s2017 from './seasons/2017.csv';
import s2018 from './seasons/2018.csv';
import s2019 from './seasons/2019.csv';
import s2020 from './seasons/2020.csv';

export const PATHS = <const>{
    '1991': s1991,
    '1992': s1992,
    '1993': s1993,
    '1994': s1994,
    '1995': s1995,
    '1996': s1996,
    '1997': s1997,
    '1998': s1998,
    '1999': s1999,
    '2000': s2000,
    '2001': s2001,
    '2002': s2002,
    '2003': s2003,
    '2004': s2004,
    '2005': s2005,
    '2006': s2006,
    '2007': s2007,
    '2008': s2008,
    '2009': s2009,
    '2010': s2010,
    '2011': s2011,
    '2012': s2012,
    '2013': s2013,
    '2014': s2014,
    '2015': s2015,
    '2016': s2016,
    '2017': s2017,
    '2018': s2018,
    '2019': s2019,
    '2020': s2020,
};
