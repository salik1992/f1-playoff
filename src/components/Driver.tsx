import styled from "styled-components";
import { Driver } from "../data";

type Props = {
    offsetIndex: number,
    driver: Driver,
}

const HEIGHT = 50;

const DriverWrap = styled.div`
    position: absolute;
    display: flex;
    top: 0;
    left: 0;
    right: 0;
    height: ${HEIGHT}px;
    line-height: ${HEIGHT}px;
    color: #ffffff;
    background-image: linear-gradient(to top, #000000, #222222);
    transition: transform 800ms;

    .name {
        flex: 1;
    }

    .position, .points, .points-from-last-race {
        padding: 0 20px;
        min-width: 20px;
        font-weight: bold;
    }

    .position {
        text-align: center;
    }

    .points, .points-from-last-race {
        text-align: right;
    }

    .points-from-last-race {
        color: #00aa00;

        &.points-0 {
            color: #999999;
        }
    }
`;

export const DriverView = ({
    offsetIndex, driver: {
        position, name, points, pointsFromLastRace,
    } }: Props) => (
    <DriverWrap style={{ transform: `translateY(${offsetIndex * HEIGHT}px)` }}>
        <div className="position">
            {position + 1}
        </div>
        <div className="name">
            {name}
        </div>
        <div className="points">
            {points}
        </div>
        <div className={`points-from-last-race points-${pointsFromLastRace}`}>
            +{pointsFromLastRace}
        </div>
    </DriverWrap>
);
