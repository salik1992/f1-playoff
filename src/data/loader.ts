import { PATHS } from './paths';
import { parseSeason } from './parser';

export const loadSeason = async (seasonYear: keyof typeof PATHS) => {
    const csv = await (await fetch(PATHS[seasonYear])).text();
    return parseSeason(csv);
};
