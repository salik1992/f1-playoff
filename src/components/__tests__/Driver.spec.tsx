import { render } from '@testing-library/react';
import { Driver } from '../../data';
import { DriverView } from '../Driver';

const DRIVER = {
    name: 'Test DRIVER',
    position: 0,
    points: 20,
    pointsFromLastRace: 10,
} as unknown as Driver;

describe('Driver', () => {
    it('should render driver', () => {
        render(<DriverView offsetIndex={0} driver={DRIVER} />);
        expect(document.body.innerHTML).toMatchSnapshot();
    });
});
