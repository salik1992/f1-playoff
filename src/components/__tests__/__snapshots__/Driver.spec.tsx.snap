// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Driver should render driver 1`] = `"<div><div style=\\"transform: translateY(0px);\\" class=\\"sc-bcXHqe hvrhnk\\"><div class=\\"position\\">1</div><div class=\\"name\\">Test DRIVER</div><div class=\\"points\\">20</div><div class=\\"points-from-last-race points-10\\">+10</div></div></div>"`;
